### PR TITLE
Laboratory work #3, Varvara Migacheva - 22FPL2

### PR DESCRIPTION
**Pull request by varmig2000**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -76,6 +97,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not (isinstance(element_id, int) and element_id in self._storage.values()):
+            return None
+        for letter, identifier in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:102]: again - filter?

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        for token in text.lower():
+            if token.isalpha():
+                tokens.append(token)
+            elif token.isspace() and tokens[-1] != self._end_of_word_token and tokens:
+                tokens.append(self._end_of_word_token)
+        if len(tokens) == 0:
+            return None
+        if not tokens[-1].isalnum():
+            tokens.append(self._end_of_word_token)
+        if text[-1] in string.punctuation:
```
[the remark this answers is not in the document]
Comment (varmig2000) [reply lab_3_generate_by_ngrams/main.py:58]: i have removed the first condition

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         """
+        if not (isinstance(text, str) and text):
+            return None
+        tokens = []
+        for token in text.lower():
+            if token.isalpha():
+                tokens.append(token)
+            elif token.isspace() and tokens[-1] != self._end_of_word_token and tokens:
+                tokens.append(self._end_of_word_token)
+        if len(tokens) == 0:
```
[the remark this answers is not in the document]
Comment (varmig2000) [reply lab_3_generate_by_ngrams/main.py:54]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not (isinstance(text, str) and text):
+            return None
+        tokens = []
+        for token in text.lower():
```
[the remark this answers is not in the document]
Comment (varmig2000) [reply lab_3_generate_by_ngrams/main.py:49]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (varmig2000) [reply lab_3_generate_by_ngrams/main.py:102]: fixed

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
please, adress all comments

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -104,6 +144,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
```
[the remark this answers is not in the document]
Comment (varmig2000) [reply lab_3_generate_by_ngrams/main.py:147]: fixed

**Pull request closed without merging**